OGM-359 Stored procedures for Mongo

DavideD opened the pull request:
[OGM-359](https://hibernate.atlassian.net/browse/OGM-359)

Follows #976 

While  I was checking the latest PR on stored procedures I noticed several issues and ended up fixing them and rebase the history. I think now it's in a state the we can merge. In particular,  this PR:

* Fix the logic by making it more similar to the one in ORM
* Removes commons-lang library to escape strings and created a utility method that hopefully will work
* Adds test cases for when an exception is raised
* NoSQLProcedureResultSetOutputImpl.java is now the same as the one in ORM. The problem with the one in ORM is that it is package private
* Clean up the exception messages, they are more user friendly now 

There are still a couple of tests to add:
* Check storeprocedures returning multiple entities
* Test case for stored procedures in MongoDB causing log.multipleDocumentReturnedByStoredProcedure

I'll create a separate issue for these latest couple of cases

DavideD commented:
@schernolyas maybe you can have a look and see if the changes I've applied make sense. Thanks

schernolyas commented:
Okey @DavideD . I will look to your version.

schernolyas commented:
@DavideD  I think ... we have to mark the feature as "experimental"

DavideD commented:
> @DavideD I think ... we have to mark the feature as "experimental"

Good point, I will

gsmet commented on core/src/main/java/org/hibernate/ogm/util/impl/StringHelper.java:
```diff
@@ -78,4 +79,31 @@ public static String join(Iterator<?> iterator, String separator) {
+	 * @param text the string to escape
+	 * @return the escaped text or {@code null} if text is {@code null}
+	 */
+	public static String escape(String text) {
```
This is not a generic escaping utility so `StringHelper.escape()` is not a good choice.

So maybe I would keep it in the MongoDB module and at least call it `escapeJsonString()`.

By the way, are you sure we need to escape `/`? I find it a bit weird. I'm also wondering if we need to escape `"`, considering the string is wrapped into `'`. It might be the case, I'm not very familiar with MongoDB escaping rules but I think we should test it to be sure we are not escaping things too much. Unless it was already tested?

gsmet commented on documentation/manual/src/main/asciidoc/modules/query.asciidoc:
```diff
@@ -8,6 +8,7 @@ With Hibernate OGM, you have a few alternatives that should get you covered:
 * Use JPQL - only for simple queries for now
 * Use the NoSQL native query mapping the result as managed entities
 * Use Hibernate Search queries - primarily full-text queries
+* Use the stored procedure mapping the result as managed entities
```
Maybe `Use stored procedures mapping the results to managed entities` would be better?

gsmet commented on documentation/manual/src/main/asciidoc/modules/query.asciidoc:
```diff
@@ -421,3 +422,73 @@ go check the https://docs.jboss.org/hibernate/search/{hibernate-search-major-min
+Often you want the raw power of the underlying NoSQL query engine.
+Even if that costs you portability.
+
+Hibernate OGM addresses that problem by letting you express stored procedures (e.g. in Server-side JavaScript for MongoDB )
```
* Let's drop the cap on `Server-side`, `server-side` it is.
 * `MongoDB )`, we should remove the space before the parenthesis

gsmet commented on documentation/manual/src/main/asciidoc/modules/query.asciidoc:
```diff
@@ -421,3 +422,73 @@ go check the https://docs.jboss.org/hibernate/search/{hibernate-search-major-min
+The first form accepts a result class if your result set maps the mapping definition of the entity.
+The second form accepts the name of a resultSetMapping
+and lets you customize how properties are mapped to columns by the query.
+You can also used a predefined named query which defines its result set mapping.
```
s/used/use/

gsmet commented on documentation/manual/src/main/asciidoc/modules/query.asciidoc:
```diff
@@ -421,3 +422,73 @@ go check the https://docs.jboss.org/hibernate/search/{hibernate-search-major-min
+
+javax.persistence.EntityManager em = ...
+
+StoredProcedureQuery storedProcedureQuery = em.createStoredProcedureQuery( "mostExpensiveCarsPerYear", Car.class );
```
It's a minor thing but could we be consistent with the naming of the stored procedure? Either use underscores or camel case for both.

gsmet commented on documentation/manual/src/main/asciidoc/modules/query.asciidoc:
```diff
@@ -421,3 +422,73 @@ go check the https://docs.jboss.org/hibernate/search/{hibernate-search-major-min
+[WARNING]
+====
+'OUT' and 'IN_OUT' parameters of stored procedures are not supported yet.
+Main reason of it is supported data storages are supports 'IN' parameters only.
```
This sentence is weird.

Maybe `The main reason of it is that supported data storages only support 'IN' parameters` would be better?

gsmet commented on mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java:
```diff
@@ -1733,6 +1739,95 @@ private static ReadPreference getReadPreference(AssociationContext associationCo
+		List<Object> positionalParameters = params.getPositionalParameters();
+		for ( Object paramValue : positionalParameters ) {
+			if ( paramValue instanceof String ) {
+				//need for escape char "'"
```
Depending on what we decide needs escaping, we should update this comment.

gsmet commented on mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/logging/impl/Log.java:
```diff
@@ -134,4 +136,12 @@
+	HibernateException dialectDoesNotSupportNamedParametersForStoredProcedures( @FormatWith(ClassObjectFormatter.class) Class<?> dialectClass );
+
+	@Message(id = 1240, value = "Procedures returning muliple documents are not supported. Procedure '%1$s' returned %2$d results")
+	HibernateException multipleDocumentReturnedByStoredProcedure(String storedProcedureName, int size);
```
We usually have a . at the end of the log messages (this remark is for all the messages above).

+ s/muliple/multiple/

gsmet replied:
@DavideD the s/muliple/multiple/ change was not done.

gsmet commented on documentation/manual/src/main/asciidoc/modules/mongodb.asciidoc:
```diff
@@ -3473,6 +3473,55 @@ List<Poem> result = session.createNativeQuery( nativeQuery ).addEntity( Poem.cla
+
+==== Server-side JavaScript and stored procedures
+
+In MongoDB, it is possible to call server side javascripts as if it is a stored procedure.
```
* In the main doc, you call it `server-side JavaScript`, let's be consistent about the case. Maybe use `server-side JavaScript scripts` here? It looks redundant but I think it's the way to go.
 * `as if they were stored procedures`

gsmet commented on documentation/manual/src/main/asciidoc/modules/mongodb.asciidoc:
```diff
@@ -3473,6 +3473,55 @@ List<Poem> result = session.createNativeQuery( nativeQuery ).addEntity( Poem.cla
+In MongoDB, it is possible to call server side javascripts as if it is a stored procedure.
+You can use the existing methods in JPA:
+
+.Calling server side javascript with Hibernate OGM with positional parameters
```
same here `server-side JavaScript`

I would drop the `with Hibernate OGM`. It's 2 `with` and, well, we are in the Hibernate OGM documentation :).

gsmet commented on documentation/manual/src/main/asciidoc/modules/mongodb.asciidoc:
```diff
@@ -3473,6 +3473,55 @@ List<Poem> result = session.createNativeQuery( nativeQuery ).addEntity( Poem.cla
+  }
+
+  EntityManager em = ...
+  StoredProcedureQuery storedProcedureQuery = em.createStoredProcedureQuery( "findMostExpensiveCars", Car.class );
```
Same here about the consistency remark below (sorry I reviewed the main doc first so the comments look a bit backward). Let's name the stored procedures either with underscores or camel case. (Mentioning it here so that it's also fixed if you choose the underscores).

gsmet commented on documentation/manual/src/main/asciidoc/modules/mongodb.asciidoc:
```diff
@@ -3473,6 +3473,55 @@ List<Poem> result = session.createNativeQuery( nativeQuery ).addEntity( Poem.cla
+----
+====
+
+This example will work assuming that there is a javascript function in MongoDB defined
```
"... there is a `findMostExpensiveCars` JavaScript function defined in MongoDB..." would be better

The word order at the end of the sentence could be improved too: "that can be mapped to the `Car` entity"

gsmet commented on documentation/manual/src/main/asciidoc/modules/mongodb.asciidoc:
```diff
@@ -3473,6 +3473,55 @@ List<Poem> result = session.createNativeQuery( nativeQuery ).addEntity( Poem.cla
+`findMostExpensiveCars` and that the result of the function is a list of cars that can be mapped
+by the entity `Car`.
+
+.Calling server side javascript with Hibernate OGM with positional parameters
```
Same remark here about the title. But maybe the title should be changed anyway as it does not fit very well to what is there?

gsmet commented on documentation/manual/src/main/asciidoc/modules/mongodb.asciidoc:
```diff
@@ -3473,6 +3473,55 @@ List<Poem> result = session.createNativeQuery( nativeQuery ).addEntity( Poem.cla
+====
+
+I will let the details about server side funtions to
+https://docs.mongodb.com/manual/core/server-side-javascript[the MongoDB reference documentation].
```
Maybe "More details about server-side JavaScript functions can be found in the MongoDB reference documentation."

We usually avoid the `I` in documentation.

DavideD commented on core/src/main/java/org/hibernate/ogm/storedprocedure/impl/NoSQLProcedureCallMemento.java:
```diff
@@ -0,0 +1,43 @@
+		return new NoSQLProcedureCallImpl( session, this );
+	}
+
+	public Map<String, Object> getHintsMap() {
```
Right, thanks

DavideD commented on core/src/main/java/org/hibernate/ogm/storedprocedure/impl/NoSQLProcedureResultSetOutputImpl.java:
```diff
@@ -0,0 +1,54 @@
+
+	@Override
+	public Object getSingleResult() {
+		final List<?> results = getResultList();
```
+1
Just, so you know, I copied and paste the class from ORM and just renamed it.

DavideD commented:
@gsmet 
I've applied the remarks and update the PR.
In the end I think we only need to escape double quotes. We can also add tests later if somebody has a use case. I also add a check to  verify if the string is a number or a text.
All this part would require more testing but I think it can be merged for now.

I also added a note in the documentation to highlight that this feature is experimental.

Cheers

gsmet commented on mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java:
```diff
@@ -1733,6 +1742,114 @@ private static ReadPreference getReadPreference(AssociationContext associationCo
+			if ( requiresQuotes( paramValue ) ) {
+				String escapedValue = StringHelper.escapeDoubleQuotesForJson( paramValue.toString() );
+				// I think we only have to worry about double quotes for String
+				commandLine.append( '"' ).append( escapedValue ).append( '"' );
```
I think you also have to escape the `\` character. Otherwise a `\"` string will be changed to `"`.

DavideD replied:
I'm ont sure I follow, but I added a test and it seems to work
Feel free to add another one before making changes, please

DavideD commented:
Merged https://github.com/hibernate/hibernate-ogm/pull/979

I've refactored the code and squashed the commits to create a clean history.
Everything test works now.

gsmet commented:
@DavideD Nice, glad to see this in.

DavideD commented:
@gsmet Yes, it definitely took longer than expected.

I also added a test to return value without mapping the result and found another couple of problems :/
